File: Java/581.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public int findUnsortedSubarray(int[] nums) {
        
		int head = 0;
		int tail = nums.length-1;
		int flag = 0;
		for(int i = 0 ; i < nums.length ; ++ i)
			if(nums[i] < nums[head])
			{
				while(head >= 0 && nums[i] < nums[head])
					head --;
				if(head < 0)
					break;
				flag = 1;
			}
			else if(flag == 0)
				head = i;
		
		flag = 0;
		for(int i = nums.length-1 ; i >= 0 ; -- i)
			if(nums[i] > nums[tail])
			{
				while(tail < nums.length && nums[i] > nums[tail])
					tail ++;
				if(tail == nums.length)
					break;
				flag = 1;
			}
			else if(flag == 0)
				tail = i;
		return Math.max(tail - head - 1 , 0 );
    }
}
__________________________________________________________________________________________________
sample 38936 kb submission
class Solution {
    public int findUnsortedSubarray(int[] nums) {
        if(nums == null || nums.length == 0){
            return -1;
        }

        int start = 0, end = 0;
        int min = nums[nums.length - 1];
        int max = nums[0];

        for(int i = 1; i < nums.length; i++){
            if(nums[i] < max){
                end = i;
            }
            max = Math.max(max, nums[i]);
        }

        for(int i = nums.length - 1; i >= 0; i--){
            if(nums[i] > min){
                start = i;
            }
            min = Math.min(min, nums[i]);
        }
        return start != end ? end - start + 1 : 0;
    }
}
__________________________________________________________________________________________________

File: Python3/581.py
__________________________________________________________________________________________________
sample 212 ms submission
class Solution:
    
    # 思路：最少要排序的长度，是从左第一个下降点，和从右第一个上升点间的距离。
    # 而不用排序又要满足条件，左边最大值小于右边所有，右边最小值大于左边所有，
    # 边界问题是最麻烦的，思路其实懂了就很简单
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return 0
        l = len(nums)
        left = 0
        while left < l-1:
            if nums[left] <= nums[left+1]:
                left += 1
            else:
                break
        right = l-1
        while right > left:
            if nums[right] >= nums[right-1]:
                right -= 1
            else:
                break
        if left >= right:
            return 0
        leftMin = min(nums[left+1:])
        rightMax = max(nums[:right])
        while left >= 0:
            if nums[left] > leftMin:
                left -= 1
            else:
                break
        while right < l:
            if nums[right] < rightMax:
                right += 1
            else:
                break
        
        return right - left - 1
__________________________________________________________________________________________________
sample 13844 kb submission
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        if nums == sorted(nums):
            return 0
        
        left = 0
        right = 0
        
        sort_arr = sorted(nums)
        for i in range(0, len(nums)):
            if nums[i] != sort_arr[i]:
                left = i
                break
        for j in range(len(nums) - 1, left, -1):
             if nums[j] != sort_arr[j]:
                right = j
                break
        
        return right - left + 1
__________________________________________________________________________________________________
